apps/emqx_node_rebalance/src/emqx_node_rebalance_evacuation.erl

```diff
@@ -0,0 +1,296 @@
+                ok ->
+                    Data = init_data(#{}, Opts),
+                    ok = warn_enabled(),
+                    {ok, evicting_conns, Data, [{state_timeout, 0, evict_conns}]};
+                {error, eviction_agent_busy} ->
+                    emqx_node_rebalance_evacuation_persist:clear(),
+                    {ok, disabled, #{}}
+            end;
+        none ->
+            {ok, disabled, #{}}
+    end.
+
+%% start
+handle_event({call, From},
+             {start, #{server_reference := ServerReference} = Opts},
+             disabled,
+             #{} = Data) ->
+    case emqx_eviction_agent:enable(?MODULE, ServerReference) of
+        ok ->
+            NewData = init_data(Data, Opts),
+            ok = emqx_node_rebalance_evacuation_persist:save(Opts),
+            ?LOG(warning, "Node evacuation started"),
+            {next_state,
+             evicting_conns,
+             NewData,
+             [{state_timeout, 0, evict_conns},
+              {reply, From, ok}]};
+        {error, eviction_agent_busy} ->
+            {keep_state_and_data,
+              [{reply, From, {error, eviction_agent_busy}}]}
+    end;
+handle_event({call, From}, {start, _Opts}, _State, #{}) ->
+    {keep_state_and_data,
+     [{reply, From, {error, already_started}}]};
+
+%% stop
+handle_event({call, From}, stop, disabled, #{}) ->
+    {keep_state_and_data,
+     [{reply, From, {error, not_started}}]};
+handle_event({call, From}, stop, _State, Data) ->
+    ok = emqx_node_rebalance_evacuation_persist:clear(),
+    _ = emqx_eviction_agent:disable(?MODULE),
+    ?LOG(warning, "Node evacuation stopped"),
+    {next_state,
+     disabled,
+     deinit(Data),
+     [{reply, From, ok}]};
+
+%% status
+handle_event({call, From}, status, disabled, #{}) ->
+    {keep_state_and_data,
+     [{reply, From, disabled}]};
+handle_event({call, From}, status, State, #{migrate_to := MigrateTo} = Data) ->
+    Stats = maps:with(
+              [initial_conns, current_conns,
+               initial_sessions, current_sessions,
+               server_reference, conn_evict_rate, sess_evict_rate],
+              Data),
+    {keep_state_and_data,
+     [{reply, From, {enabled, Stats#{state => State, migrate_to => migrate_to(MigrateTo)}}}]};
+
+%% conn eviction
+handle_event(state_timeout,
+             evict_conns,
+             evicting_conns,
+             #{conn_evict_rate := ConnEvictRate,
+               wait_takeover := WaitTakeover} = Data) ->
+    case emqx_eviction_agent:status() of
+        {enabled, #{connections := Conns}} when Conns > 0 ->
+            ok = emqx_eviction_agent:evict_connections(ConnEvictRate),
+            ?tp(debug, node_evacuation_evict_conn, #{conn_evict_rate => ConnEvictRate}),
+            ?LOG(warning, "Node evacuation evict_conns, count=~p, conn_evict_rate=~p",
+                 [Conns, ConnEvictRate]),
+            NewData = Data#{current_conns => Conns},
```

Comment:
Q: the true current connection count is lower than this because some connections were just evicted?